### PR TITLE
Avoid slow type checks against promises on outbound buffer's progress

### PR DESCRIPTION
**Pull request by franz1981**
Motivation:

ChannelOutboundBuffer's progress perform expensive type checks that can reduce scalability or just performing wasteful operations over common Netty types. This behaviour pop-up in few real-cases, but sadly it cannot be captured easily by java profilers due to https://github.com/jvm-profiling-tools/async-profiler/issues/673

Modification:

Prefer direct class checks for void promises and guarding Netty types by using concrete class type checks over interface type checks, that will be left to user-defined promises instead.

Result:

Faster type checks for Netty types and faster outbound buffer progress's performance.

### CODE REVIEW

**Comment by franz1981** [comment]
@vietj as well: I have no idea if vert-x does impl it's own promises that can hit the slowest path here (the interface type check)

**Comment by franz1981** [comment]
@normanmaurer do you know how frequently can happen for promises types to hit the very last type check and fail (because not ChannelProgressivePromise`?)  i.e.
https://github.com/netty/netty/blob/404580c31ca80a43b321af9279f4e0d079cc1059/transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java#L257-L259

I don't want to harm any other possible case here, but, as reported in the comment, I'm ofteen seeing this pop-up in real world assembly profiling, due to the super-slow path check of void promises (mostly).

**Comment by franz1981** [comment]
This is a part (I've added some notes to help the reader) related this issue; in this case was happening with void promises (very common hot path on Netty 4.1):
```asm
           │  0x00007f1260410945: mov    0x20(%rsi),%rcx     ; reading it the secondary super cache field of VoidChannelPromise's Klass into rcx		
  0.02%    │  0x00007f1260410949: movabs $0x80011abe0,%rdi   ;   {metadata(&apos;io/netty/channel/ChannelProgressivePromise&apos;)}
           │  0x00007f1260410953: cmp    %rdi,%rcx	     ; comparing it against ChannelProgressivePromise (and failing, because ssc contains only implemented types	
           │  0x00007f1260410956: je     0x00007f1260410a84  ; never happen here  
  0.02%    │  0x00007f126041095c: movabs $0x80011abe0,%rax   ;   {metadata(&apos;io/netty/channel/ChannelProgressivePromise&apos;)}
           │  0x00007f1260410966: push   %rax		     ; breaking java profiler by modifying the stack frame size :"( by pushing on the stack ChannelProgressivePromise 	
  0.00%    │  0x00007f1260410967: mov    %rax,%rax
           │  0x00007f126041096a: mov    0x28(%rsi),%rdi     ; reading the secondary supers field of VoidChannelPromise's Klass
  0.03%    │  0x00007f126041096e: mov    (%rdi),%ecx
  0.02%    │  0x00007f1260410970: add    $0x8,%rdi
  0.00%    │  0x00007f1260410974: test   %rax,%rax
  0.18%    │  0x00007f1260410977: repnz scas %es:(%rdi),%rax ; perform O(N) search on secondary supers field of VoidChannelPromise to find ChannelProgressivePromise  
  0.81%    │  0x00007f126041097a: pop    %rax	             ; skidding here: this poor instruction took all the blame of the CPU cycle's cost :"(
  0.21%   ╭│  0x00007f126041097b: jne    0x00007f1260410985
          ││  0x00007f1260410981: mov    %rax,0x20(%rsi)     ; saving ChannelProgressivePromise into ssc's of VoidChannelPromise: this cannot happen, is not an implemented type
  0.00%   ↘│  0x00007f1260410985: je     0x00007f1260410a84  ;*instanceof {reexecute=0 rethrow=0 return_oop=0}
           │                                                 ; - io.netty.channel.ChannelOutboundBuffer::progress@45 (line 246)
           │                                                 ; - io.netty.channel.ChannelOutboundBuffer::removeBytes@73 (line 349)
           │                                                 ; - io.netty.channel.epoll.AbstractEpollStreamChannel::writeBytesMultiple@87 (line 305)
           │                                                 ; - io.netty.channel.epoll.AbstractEpollStreamChannel::doWriteMultiple@45 (line 510)
           │                                                 ; - io.netty.channel.epoll.AbstractEpollStreamChannel::doWrite@31 (line 422)
```

File: transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java [hunk 1]
```diff
@@ -243,7 +243,18 @@ public void progress(long amount) {
         ChannelPromise p = e.promise;
         long progress = e.progress + amount;
         e.progress = progress;
-        if (p instanceof ChannelProgressivePromise) {
+        if (p == null) {
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java:246]: This can never happen

**Review by normanmaurer — CHANGES_REQUESTED** [review]

**Review by chrisvest — APPROVED** [review]

**Comment by franz1981** [comment]
Ready to go @normanmaurer !

**Comment by normanmaurer** [comment]
Thanks a lot!

**Pull request merged**